/**
 * This is a module containing control for accessories on the platform:
 * Buttons, lights, turn signals, etc.
 *
 */

syntax = "proto3";

/*
 * Car accessory module contains controls for generic car accessories
 */
package CarAccessoryModule;
option go_package = "../internal/pkg/ba_proto;ba_proto";
option csharp_namespace = "Google.Protobuf.ba_proto";


/**
 * Car accessory status information, contains status message of one of supported device
 */
message CarAccessoryStatus {
  /*
   * Information about source device
   */
  Device device = 1;

  /*
   * Device specific status message
   */
  oneof DeviceStatusType{
    ButtonStatus buttonStatus = 2;
  }
}

/**
 * Car accessory command information, contains command message of one of supported device
 */
message CarAccessoryCommand {
  /*
   * Information about target device
   */
  Device device = 1;

  /*
   * Device specific command message
   */
  oneof DeviceCommandType{
    ButtonCommand buttonCommand = 2;
  }
}

/**
 * Car accessory error information, contains error message of one of supported device
 */
message CarAccessoryError {
  /*
   * Information about source device
   */
  Device device = 1;

  /*
   * Device specific error message
   */
  oneof DeviceErrorType{
    ButtonError buttonError = 2;
  }
}

/*
 * Message contains information about device
 */
message Device{
  enum Module {
    RESERVED_MODULE = 0;
    CAR_ACCESSORY_MODULE = 2;
  }
  /*
   * Enum contains all device types supported by module, when module client is connecting, deviceType value should be set according to this enum
   */
  enum DeviceType{
    BUTTON = 0;
  }

  Module module = 1;

  DeviceType deviceType = 2;

  /*
   * Unique device name
   */
  string deviceName = 3;
}

message ButtonStatus{
  bool isPressed = 1;
}

message ButtonCommand{
  enum Command{
    PRESS = 0;
  }
  Command command = 1;
}

message ButtonError{
  uint32 pressCount = 1;
}